scheduler/scheduler.go

```diff
@@ -220,10 +225,154 @@ func spawnHosts(ctx context.Context, newHostsNeeded map[string]int) (map[string]
 		"span":      time.Since(startTime).String(),
 		"duration":  time.Since(startTime),
 	})
-
 	return hostsSpawnedPerDistro, nil
 }
 
+// generateHostOptions generates host options based on what kind of host it is:
+// regular host or container
+func generateHostOptions(d distro.Distro) (cloud.HostOptions, error) {
+	if d.MaxContainers != 0 {
+		parent, err := findAvailableParent(d)
+		if err != nil {
+			return cloud.HostOptions{}, err
+		}
+		hostOptions := cloud.HostOptions{
+			ParentID: parent.Id,
+			UserName: evergreen.User,
+			UserHost: false,
+		}
+		return hostOptions, nil
+	}
+
+	hostOptions := cloud.HostOptions{
+		UserName: evergreen.User,
+		UserHost: false,
+	}
+	return hostOptions, nil
+}
+
+// generateParentHostOptions generates host options for a parent host
+func generateParentHostOptions() cloud.HostOptions {
+	return cloud.HostOptions{
+		HasContainers: true,
+		UserName:      evergreen.User,
+		UserHost:      false,
+	}
+}
+
+// FindAvailableParent finds a parent host that can accommodate container
+// based on oldest parent
+func findAvailableParent(d distro.Distro) (host.Host, error) {
+	allParents, err := host.FindAllRunningParents()
+	if err != nil {
+		return host.Host{}, err
+	}
+
+	oldestParent := host.Host{}
+	oldestTime := time.Now().Add(time.Hour * 1)
+	for _, parent := range allParents {
+		currentContainers, err := parent.GetContainers()
+		if err != nil {
+			return host.Host{}, err
+		}
+		if parent.CreationTime.Before(oldestTime) && len(currentContainers) < d.MaxContainers {
+			oldestParent = parent
+			oldestTime = parent.CreationTime
+		}
+	}
+	if oldestParent.Id == "" {
+		return host.Host{}, errors.New("no available parent found for container")
+	}
+	return oldestParent, nil
+
+}
+
+// CalcNewParentsNeeded returns the number of additional parents needed to
+// accommodate new containers
+func calcNewParentsNeeded(numCurrentParents, numContainersNeeded, numExistingContainers int, d distro.Distro) int {
+	if numCurrentParents*d.MaxContainers <= numExistingContainers+numContainersNeeded {
+		return int(math.Ceil(float64(numContainersNeeded) / float64(d.MaxContainers)))
+	}
+	return 0
+}
+
+// parentCapacity calculates number of new parents and new containers to create
+// checks to make sure we do not create more parents than allowed
+func parentCapacity(d distro.Distro, numNewParents, numCurrentParents, numCurrentContainers, numContainersToSpawn int) (int, int) {
```

Comment:
See comment above. I think this function need only return parents needed, not containers.